includes/class-sensei-frontend.php

```diff
@@ -1237,10 +1237,9 @@ public function sensei_course_start() {
 	 * @return bool True if successful.
 	 */
 	public function manually_enrol_learner( $user_id, $course_id ) {
-		$enrolment_manager = Sensei_Course_Enrolment_Manager::instance();
-		$manual_enrolment  = $enrolment_manager->get_manual_enrolment_provider();
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
 
-		return $manual_enrolment && $manual_enrolment->enrol_learner( $user_id, $course_id );
+		return $course_enrolment->enrol( $user_id );
```

Comment:
I wonder if this could lead to some inconsistency. If the learner is removed, this method just restores them (eg removes them from the 'removed' list), but won't set the manual enrolment as a provider. That could mean that when their original provider changes something, for example, a subscription expires, it can take away the enrolment, even though the user should be still enrolled thought the manual provider.

WDYT of calling `Sensei_Course_Enrolment::restore_learner` directly, to clear them from the removed list, then going ahead with the new manual enrolment? Could be part of `Sensei_Course_Manual_Enrolment_Provider::enrol_learner` too.

cc @renatho you might know these parts better :)

Reply:
> If the learner is removed, this method just restores them (eg removes them from the 'removed' list), but won't set the manual enrolment as a provider

Hmm, unless I'm looking at it from the wrong perspective, it looks like `Sensei_Course_Enrolment::enrol` both restores the learner if it's removed ([source](https://github.com/Automattic/sensei/blob/a956e2a3e6ed6490bf60edc4462f37dea73ae237/includes/enrolment/class-sensei-course-enrolment.php#L598-L605)), **and** uses the manual provider to enrol it ([source](https://github.com/Automattic/sensei/blob/a956e2a3e6ed6490bf60edc4462f37dea73ae237/includes/enrolment/class-sensei-course-enrolment.php#L608-L615)), as [this method did previously](https://github.com/Automattic/sensei/blob/a956e2a3e6ed6490bf60edc4462f37dea73ae237/includes/class-sensei-frontend.php#L1240-L1243).

Am I missing something here?

> WDYT of calling Sensei_Course_Enrolment::restore_learner directly, to clear them from the removed list, then going ahead with the new manual enrolment? Could be part of Sensei_Course_Manual_Enrolment_Provider::enrol_learner too.

I thought about doing that, but in the end, it looked weird given that `Sensei_Course_Enrolment` already has some "default" integration with `Sensei_Course_Manual_Enrolment_Provider`, and vice-versa, so just using  `Sensei_Course_Enrolment::enrol` looked better in this case.

Like I explained in the PR description, honestly unless I forgetting something else here the main issue with calling that method is the [`sensei_admin_enrol_user` action that it calls](https://github.com/Automattic/sensei/blob/a956e2a3e6ed6490bf60edc4462f37dea73ae237/includes/enrolment/class-sensei-course-enrolment.php#L596). I mean, that's not an issue for SWCPC specifically (which doesn't use this method that I changed in this PR at all), but it *might* affect other plugins/use-cases.

Reply:
I'm referring to this check:
https://github.com/Automattic/sensei/blob/a956e2a3e6ed6490bf60edc4462f37dea73ae237/includes/enrolment/class-sensei-course-enrolment.php#L603-L605
If, after restoring them, they have enrolment from a different provider, it won't do the manual enrolment.

Reply:
Ahhhhh, fair enough! I see what you`re talking here. 

However, about that, I feel like it's more of...a duplicated check.

See, the method that I changed, `manually_enrol_learner`, is only called [inside the method `sensei_course_start`](https://github.com/Automattic/sensei/blob/810ed0606c63d38351d99a82512251fdb9922cf3/includes/class-sensei-frontend.php#L1194-L1198), [which checks if the user can manually enrol before trying to enrol it](https://github.com/Automattic/sensei/blob/810ed0606c63d38351d99a82512251fdb9922cf3/includes/class-sensei-frontend.php#L1177).

Inside the method `Sensei_Course::can_current_user_manually_enrol`, [you can see that it use the method  `Sensei_Course::is_user_enrolled` to check if the user is already enrolled](https://github.com/Automattic/sensei/blob/1f2b1776cbeb74007b8f14f3a3b3d488f18b226b/includes/class-sensei-course.php#L3227), and, inside that method, it [uses the same `Sensei_Course_Enrolment::is_enrolled` method to check if the user is in fact enrolled](https://github.com/Automattic/sensei/blob/1f2b1776cbeb74007b8f14f3a3b3d488f18b226b/includes/class-sensei-course.php#L270-L272).

So I'm pretty sure (not 100% tho, I appreciate checking it!) that this condition should always be `false`.

What do you think?

Reply:
Makes sense! (Though it can actually be [changed via a filter](https://github.com/Automattic/sensei/blob/fde6be83a05465633ec07bbe301bc7e91fbdddba/includes/class-sensei-course.php#L3245) :)) At this point then the concern is just a hypothetical edge case so fine to deal with it later if a bug report comes.

Reply:
Hey there! Fair enough, I decided to just "abandon" the risk of having conflicts with all those additional filters and actions and implemented a more traditional approach. The changes are in the commit [a0899ec](https://github.com/Automattic/sensei/commit/a0899ec1106c8b921ce9004e77129d7bf78a147e).